onnxruntime/core/graph/graph_utils.cc

```diff
@@ -191,51 +193,121 @@ static void UpdateImplicitInputNameInSubgraph(Node& node,
-  const NodeIndex output_edge_node_index = output_edge_node.Index();
-  const int output_edge_dst_arg = output_edge.GetDstArgIndex();
+  return output_edges;
+}
 
-  // check if the output from the node to remove is implicit input to the downstream node.
-  // if so, it's used in subgraph in that node and we have to check if it's valid to update that subgraph
-  if (OutputEdgeProvidesImplicitInput(output_edge)) {
-    // the output we need to wire to the downstream node, and the output name from the node we want to remove
-    const auto& output_name = GetNodeOutputName(input_edge_node, output_edge.GetSrcArgIndex());
-    const auto& removed_output_name = GetNodeOutputName(node, 0);
+/** Remove a node with a single incoming node. */
+static bool RemoveNodeWithSingleNodeIn(Graph& graph, Node& node) {
+  // Store info for and remove the output edges.
+  std::vector<GraphEdge> output_edges = RemoveAndReturnOutputEdges(graph, node);
 
-    Node& mutable_output_edge_node = *graph.GetNode(output_edge_node_index);
+  // Store info for the input edge.
+  const Node::EdgeEnd& input_edge_end = *node.InputEdgesBegin();
+  const auto& input_edge_node = input_edge_end.GetNode();
+  const GraphEdge input_edge{input_edge_node.Index(),
+                             node.Index(),
+                             input_edge_end.GetSrcArgIndex(),
+                             input_edge_end.GetDstArgIndex(),
+                             GetNodeOutputName(input_edge_node, input_edge_end.GetSrcArgIndex())};
 
-    if (CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name)) {
-      UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name);
-    } else {
-      // we can't safely remove this node
+  // Remove node (this will remove the input edge too).
+  graph.RemoveNode(node.Index());
+
+  // Create connections between the incoming node and the outgoing nodes of the node that we removed.
+  for (const auto& output_edge : output_edges) {
+    // Take care of subgraph inputs.
+    if (!CheckAndUpdateImplicitInputInSubgraph(graph, input_edge.arg_name, output_edge)) {
       return false;
     }
+
+    // Add new edge connecting the input with the output nodes directly.
+    graph.AddEdge(input_edge.src_node, output_edge.dst_node, input_edge.src_arg_index, output_edge.dst_arg_index);
   }
 
-  // Remove output edge.
-  graph.RemoveEdge(node.Index(), output_edge_node_index, output_edge.GetSrcArgIndex(), output_edge_dst_arg);
+  return true;
+}
 
-  // Remove node (this will remove the input edge too).
+/** Remove a node with a single incoming initializer (and no other incoming node). */
+static bool RemoveNodeWithSingleInitializerIn(Graph& graph, Node& node) {
+  // Store info for and remove the output edges.
+  std::vector<GraphEdge> output_edges = RemoveAndReturnOutputEdges(graph, node);
+
+  // Get incoming initializer node arg.
+  NodeArg* input_def = node.MutableInputDefs()[0];
+
```

Comment:
Do we need this to be mutable?

Reply:
I have to do {{output_node->MutableInputDefs()[output_edge.dst_arg_index] = input_def}} after, and I cannot do that if input_def is const right?

Reply:
operator= takes a const reference as an argument, so input_def can be const

---
In reply to: [271013862](https://github.com/Microsoft/onnxruntime/pull/743#discussion_r271013862) [](ancestors = 271013862)

Reply:
Resolving this -- as discussed offline, the left part returns a reference to a pointer, so I need to get the mutable input_def to be able to assign it.